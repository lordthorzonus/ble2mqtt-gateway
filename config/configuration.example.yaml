log_level: info
decimal_precision: 1
mqtt:
    host: mqtt.lan.fi
    port: 1883
    username: ble2mqtt
    password: secret
    client_id: ble2mqtt
    protocol: mqtt
gateways:
    base_topic: ble2mqtt
    ruuvitag:
        allow_unknown: false
        timeout: 60000
        devices:
            - name: Fridge Ruuvitag
              id: da21045d81a8
            - name: Balcony Second
              id: 641cae0910b5
              timeout: 10000
    miflora:
        devices:
            - name: My super plant
              id: c47c8d6e07cf
              timeout: 40000
homeassistant:
    discovery_topic: homeassistant
