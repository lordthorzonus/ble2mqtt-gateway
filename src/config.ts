import { load } from "js-yaml";
import * as fs from "fs";
import { z, ZodError } from "zod";

const mqttSchema = z.object({
    host: z.string(),
    port: z.number().int().default(1883),
    username: z.string(),
    password: z.string(),
    client_id: z.string(),
    protocol: z.union([z.literal("mqtt"), z.literal("mqtts"), z.literal("tcp"), z.literal("ssl")]).default("mqtt"),
});

const ruuvitagSchema = z
    .object({
        allow_unknown: z.boolean().default(false),
        timeout: z.number().int().default(30000),
        devices: z
            .array(
                z.object({
                    name: z.string(),
                    id: z.string(),
                    timeout: z.number().int().optional(),
                })
            )
            .min(0),
    })
    .optional();

const miFloraSchema = z
    .object({
        timeout: z.number().int().default(60000),
        devices: z
            .array(
                z.object({
                    name: z.string(),
                    id: z.string(),
                    timeout: z.number().int().optional(),
                })
            )
            .min(0),
    })
    .optional();

const configSchema = z.object({
    log_level: z.union([z.literal("info"), z.literal("debug"), z.literal("error")]).default("info"),
    mqtt: mqttSchema,
    gateways: z.object({
        base_topic: z.string().default("ble2mqtt"),
        ruuvitag: ruuvitagSchema,
        miflora: miFloraSchema,
    }),
    homeassistant: z
        .object({
            discovery_topic: z.string().default("homeassistant"),
        })
        .default({
            discovery_topic: "homeassistant",
        }),
});

const produceErrorMessage = (error: ZodError) => {
    const errorMessage = "Found following errors in configuration: \n";
    return error.issues.reduce((acc, issue) => {
        const errorPath = issue.path.join(".");

        return acc.concat(errorPath, ": ", issue.message, "\n");
    }, errorMessage);
};

const configurationFileLocation = process.env.CONFIG_FILE_LOCATION ?? `${__dirname}/../config/configuration.yaml`;
const configurationFileContent = fs.readFileSync(configurationFileLocation, "utf-8");
const config = load(configurationFileContent);

export type Config = z.infer<typeof configSchema>;

export const getConfiguration = (): Config => {
    try {
        return configSchema.parse(config);
    } catch (e) {
        if (e instanceof ZodError) {
            console.error(produceErrorMessage(e));
        }

        throw e;
    }
};
