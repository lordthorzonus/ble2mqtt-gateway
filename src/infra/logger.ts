import { createLogger, format, transports } from "winston";
import { Effect } from "effect";
import { Config } from "../config";

export type LoggerInterface = ReturnType<typeof createLogger>;
export class Logger extends Effect.Service<Logger>()("Logger", {
    effect: Effect.gen(function* () {
        const config = yield* Config;

        return createLogger({
            level: config.log_level,
            format: format.combine(
                format.timestamp({
                    format: "YYYY-MM-DD HH:mm:ss",
                }),
                format.errors({ stack: true }),
                format.splat(),
                format.json()
            ),
            defaultMeta: {
                gateway: config.gateways.base_topic,
            },
            transports: [new transports.Console()],
        });
    }),
    dependencies: [Config.Default],
}) {}
