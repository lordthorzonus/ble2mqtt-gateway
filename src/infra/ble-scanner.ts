import * as noble from "@abandonware/noble";
import { Peripheral } from "@abandonware/noble";
import { Observable } from "rxjs";
import { logger } from "./logger";

const startScanning = () => {
    logger.info("Starting scanning BLE devices");
    noble.startScanning([], true);
};

export const scan = (): Observable<Peripheral> => {
    return new Observable<Peripheral>((subscriber) => {
        if (noble.state === "poweredOn") {
            startScanning();
        }

        noble.on("discover", (peripheral: Peripheral) => {
            logger.debug("Received BLE advertisement %s", peripheral);
            subscriber.next(peripheral);
        });

        noble.on("stateChange", (state) => {
            logger.info("Noble state changed to %s", state);

            if (state === "poweredOn") {
                startScanning();
            }
        });

        return () => {
            noble.stopScanning();
            noble.removeAllListeners();
        };
    });
};
