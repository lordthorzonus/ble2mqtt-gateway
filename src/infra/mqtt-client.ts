import * as mqtt from "mqtt";

import { MqttMessage } from "../types";
import { Data, Effect } from "effect";
import { Logger } from "./logger";
import { Config } from "../config";

const mqttClient = Effect.gen(function* () {
    const config = yield* Config;
    const logger = yield* Logger;

    const client = mqtt.connect({
        clientId: config.mqtt.client_id,
        username: config.mqtt.username,
        password: config.mqtt.password,
        port: config.mqtt.port,
        protocol: config.mqtt.protocol,
        rejectUnauthorized: true,
        hostname: config.mqtt.host,
        protocolVersion: 5,
        connectTimeout: 50,
    });
    client.on("connect", () => {
        logger.info("MQTT client connected");
    });

    client.on("disconnect", () => {
        logger.info("MQTT client disconnected");
    });

    client.on("error", (error) => {
        logger.error("MQTT client error", { error });
    });

    yield* Effect.addFinalizer(() => Effect.sync(() => client.end()));

    return client;
});

export class MqttClient extends Effect.Service<MqttClient>()("MqttClient", {
    scoped: mqttClient,
    dependencies: [Config.Default, Logger.Default],
}) {}

export class MqttClientError extends Data.TaggedError("MqttClientError")<{
    message: string;
    mqttMessage: MqttMessage;
    cause?: Error;
}> {}

export const publish = (message: MqttMessage): Effect.Effect<void, MqttClientError, MqttClient | Logger> =>
    Effect.gen(function* () {
        const client = yield* MqttClient;
        const logger = yield* Logger;

        logger.debug("Publishing MQTT Message", { message });
        return yield* Effect.async<undefined, MqttClientError, MqttClient>((resume) => {
            client.publish(
                message.topic,
                message.payload,
                {
                    retain: message.retain,
                },
                (error) => {
                    if (error) {
                        resume(
                            new MqttClientError({
                                message: "Failed to publish MQTT message",
                                mqttMessage: message,
                                cause: error,
                            })
                        );
                    }
                    resume(Effect.succeed(undefined));
                }
            );
        });
    });
