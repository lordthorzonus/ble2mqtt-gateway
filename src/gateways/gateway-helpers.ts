import { Peripheral } from "@abandonware/noble";
import { DeviceAvailabilityMessage, DeviceMessage, DeviceSensorMessage, DeviceType } from "../types";
import { generateAvailabilityMessage } from "./message-generators";
import { DeviceRegistry, DeviceRegistryEntry } from "./device-registry";
import { Data, Effect, Stream, Context, Option, pipe, Schedule } from "effect";
import { Logger } from "../infra/logger";
import { Config } from "../config";

export class DeviceNotFoundError extends Data.TaggedError("DeviceNotFoundError")<{
    id: string;
}> {}

export class DeviceRegistryService extends Context.Tag("DeviceRegistryService")<
    DeviceRegistryService,
    DeviceRegistry
>() {}

const shouldDeviceAvailabilityBeBroadcast = (deviceRegistryEntry: DeviceRegistryEntry) =>
    deviceRegistryEntry.availability !== deviceRegistryEntry.lastPublishedAvailability;

const handleDeviceAvailability = (
    peripheral: Peripheral,
    deviceType: DeviceType
): Effect.Effect<Option.Option<DeviceAvailabilityMessage>, DeviceNotFoundError, DeviceRegistryService> =>
    Effect.gen(function* () {
        const id = peripheral.uuid;
        const deviceRegistry = yield* DeviceRegistryService;

        if (!deviceRegistry.has(id)) {
            deviceRegistry.registerUnknownDevice(peripheral, deviceType);
        }

        deviceRegistry.registerFoundAdvertisement(id);
        const device = deviceRegistry.get(id);

        if (device === null) {
            return yield* Effect.fail(new DeviceNotFoundError({ id }));
        }

        if (shouldDeviceAvailabilityBeBroadcast(device)) {
            deviceRegistry.registerDeviceStatusPublished(id);
            return Option.some(generateAvailabilityMessage(device, "online", peripheral));
        }

        return Option.none();
    });

export const handleBleAdvertisement = <TError, TPeripheral extends Peripheral>(
    peripheral: TPeripheral,
    deviceType: DeviceType,
    unknownDevicesAllowed: boolean,
    handleDeviceSensorData: (
        peripheral: TPeripheral,
        deviceRegistryEntry: DeviceRegistryEntry
    ) => Effect.Effect<Option.Option<DeviceSensorMessage>, TError, DeviceRegistryService | Logger>
): Effect.Effect<Iterable<DeviceMessage>, DeviceNotFoundError | TError, DeviceRegistryService | Logger> =>
    Effect.gen(function* () {
        const id = peripheral.uuid;
        const deviceRegistry = yield* DeviceRegistryService;

        if (deviceRegistry.has(id) || unknownDevicesAllowed) {
            const availabilityMessage = yield* handleDeviceAvailability(peripheral, deviceType);
            const deviceRegistryEntry = deviceRegistry.get(id);

            if (deviceRegistryEntry === null) {
                return yield* new DeviceNotFoundError({ id });
            }

            const sensorDataMessage = yield* handleDeviceSensorData(peripheral, deviceRegistryEntry);
            const messages: DeviceMessage[] = [];

            if (Option.isSome(availabilityMessage)) {
                messages.push(availabilityMessage.value);
            }

            if (Option.isSome(sensorDataMessage)) {
                messages.push(sensorDataMessage.value);
            }

            return messages;
        }

        return [];
    });

export const streamUnavailableDevices: Stream.Stream<DeviceAvailabilityMessage, never, DeviceRegistryService | Config> =
    pipe(
        Effect.gen(function* () {
            const deviceRegistry = yield* DeviceRegistryService;
            const config = yield* Config;

            return pipe(
                Stream.fromSchedule(Schedule.spaced(config.unavailable_devices_check_interval_ms)),
                Stream.flatMap(() => {
                    return Stream.fromIterable(deviceRegistry.getUnavailableDevices());
                }),
                Stream.tap((device) =>
                    Effect.sync(() => {
                        deviceRegistry.registerDeviceStatusPublished(device.device.id);
                    })
                ),
                Stream.map((device) => generateAvailabilityMessage(device, "offline"))
            );
        }),
        Stream.fromEffect,
        Stream.flatten()
    );
