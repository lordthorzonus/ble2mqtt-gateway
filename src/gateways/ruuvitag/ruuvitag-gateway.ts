import { DeviceRegistry } from "../device-registry";
import { Peripheral } from "@abandonware/noble";
import { DeviceType } from "../../types";
import { transformPeripheralAdvertisementToSensorDataDeviceMessage } from "./ruuvitag-measurement-transformer";
import { RuuviTagGatewayConfiguration } from "../../config";
import { handleBleAdvertisement } from "../gateway-helpers";
import { Data, Effect } from "effect";
import { GatewayError, MapMessage } from "../ble-gateway";

export interface PeripheralWithManufacturerData extends Peripheral {
    advertisement: Omit<Peripheral["advertisement"], "manufacturerData"> & {
        manufacturerData: Buffer;
    };
}

export class PeripheralWithoutManufacturerDataError extends Data.TaggedError("PeripheralWithoutManufacturerDataError")<{
    peripheral: Peripheral;
}> {}

const validatePeripheral = (
    peripheral: Peripheral
): Effect.Effect<PeripheralWithManufacturerData, PeripheralWithoutManufacturerDataError> => {
    if (peripheral.advertisement.manufacturerData === undefined) {
        return new PeripheralWithoutManufacturerDataError({ peripheral });
    }

    return Effect.succeed(peripheral as PeripheralWithManufacturerData);
};

export const makeRuuvitagDeviceRegistry = (
    ruuviTagSettings: NonNullable<RuuviTagGatewayConfiguration>,
    globalDefaults: {
        defaultDecimalPrecision: number;
    }
): DeviceRegistry => {
    const deviceSettings = ruuviTagSettings.devices.map((tag) => ({
        device: {
            id: tag.id,
            type: DeviceType.Ruuvitag,
            friendlyName: tag.name,
        },
        timeout: tag.timeout,
    }));
    const defaultDecimalPrecision = ruuviTagSettings.decimal_precision ?? globalDefaults.defaultDecimalPrecision;

    return new DeviceRegistry(deviceSettings, ruuviTagSettings.timeout, defaultDecimalPrecision);
};

export const makeRuuvitagGateway =
    (ruuviTagSettings: NonNullable<RuuviTagGatewayConfiguration>): MapMessage =>
    (peripheral) =>
        Effect.flatMap(validatePeripheral(peripheral), (p) =>
            handleBleAdvertisement(
                p,
                DeviceType.Ruuvitag,
                ruuviTagSettings.allow_unknown,
                transformPeripheralAdvertisementToSensorDataDeviceMessage
            )
        ).pipe(
            Effect.catchTags({
                PeripheralWithoutManufacturerDataError: (error) =>
                    new GatewayError({
                        peripheral: error.peripheral,
                        message: `Peripheral without manufacturer data: ${peripheral.uuid}`,
                    }),
                DeviceNotFoundError: (error) =>
                    new GatewayError({
                        peripheral,
                        message: `Device not found: ${error.id}`,
                    }),
                NotValidRuuviManufacturerIdError: (error) =>
                    new GatewayError({
                        peripheral,
                        message: `Invalid Ruuvi manufacturer ID: ${error.manufacturerId}`,
                    }),
                UnsupportedDataFormatError: (error) =>
                    new GatewayError({
                        peripheral,
                        message: `Unsupported Ruuvi data format: ${error.dataFormat}`,
                    }),
            })
        );
