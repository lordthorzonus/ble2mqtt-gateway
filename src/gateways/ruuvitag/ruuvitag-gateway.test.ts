import { Effect, Layer } from "effect";
import { makeRuuvitagDeviceRegistry, makeRuuvitagGateway } from "./ruuvitag-gateway";
import { Settings } from "luxon";
import { v4 as uuid } from "uuid";
import { Peripheral } from "@abandonware/noble";
import { Logger } from "../../infra/logger";
import { Config, GlobalConfiguration } from "../../config";
import { DeviceRegistryService } from "../gateway-helpers";
import { getTestConfig } from "../../test/test-context";

jest.mock("uuid");
const mockedUuid = uuid as jest.Mock;
const defaultTimeout = 30000;

const validGatewayConfiguration: GlobalConfiguration["gateways"] = {
    base_topic: "test",
    ruuvitag: {
        timeout: defaultTimeout,
        allow_unknown: false,
        devices: [
            {
                id: "a1:b2",
                name: "my_ruuvitag",
            },
        ],
    },
};

const ruuviTagSettings = validGatewayConfiguration.ruuvitag;

if (!ruuviTagSettings) {
    throw new Error("RuuviTag configuration is required for this test");
}

const TestLayer = Layer.provideMerge(
    Logger.DefaultWithoutDependencies,
    Layer.succeed(
        Config,
        getTestConfig({
            gateways: validGatewayConfiguration,
        })
    )
);

describe("RuuviTag Gateway", () => {
    const originalNow = Settings.now;
    const peripheral = {
        advertisement: {
            manufacturerData: Buffer.from("99040512FC5394C37C0004FFFC040CAC364200CDCBB8334C884F", "hex"),
        },
        address: "a1:b2",
        rssi: 23,
        uuid: "a1:b2",
    } as Peripheral;

    beforeEach(() => {
        Settings.now = () => new Date("2019-10-10T00:00:00.000Z").valueOf();
        Settings.defaultZone = "UTC";
        mockedUuid.mockReturnValue("mock-uuid");
    });

    afterEach(() => {
        Settings.now = originalNow;
        jest.resetAllMocks();
    });

    const testWithFreshDeviceRegistry = <A, E, R extends Config | Logger | DeviceRegistryService>(
        program: Effect.Effect<A, E, R>
    ) => {
        const deviceRegistry = makeRuuvitagDeviceRegistry(ruuviTagSettings, {
            defaultDecimalPrecision: 2,
        });

        const TestLayerWithDeviceRegistry = Layer.merge(
            TestLayer,
            Layer.succeed(DeviceRegistryService, deviceRegistry)
        );

        return Effect.provide(
            Effect.gen(function* () {
                return yield* program;
            }),
            TestLayerWithDeviceRegistry
        );
    };

    describe("ruuvitagGateway()", () => {
        it("should produce a availability and a payload message for a found ruuvitag advertisement", async () => {
            const program = Effect.gen(function* () {
                const gateway = makeRuuvitagGateway(ruuviTagSettings);
                const messages = yield* gateway(peripheral);
                return Array.from(messages);
            });

            const messages = await Effect.runPromise(testWithFreshDeviceRegistry(program));
            expect(messages).toMatchSnapshot();
        });

        it("should complete without emitting messages if unknown ruuvitag advertisement is received", async () => {
            const program = Effect.gen(function* () {
                const gateway = makeRuuvitagGateway(ruuviTagSettings);
                const messages = yield* gateway({
                    ...peripheral,
                    uuid: "no-such-ruuvitag",
                } as Peripheral);
                return Array.from(messages);
            });

            const messages = await Effect.runPromise(testWithFreshDeviceRegistry(program));
            expect(messages).toHaveLength(0);
        });

        it("should not emit the device message until all sensor events are received once", async () =>
            Effect.runPromise(
                testWithFreshDeviceRegistry(
                    Effect.gen(function* () {
                        const gateway = makeRuuvitagGateway(ruuviTagSettings);

                        // First message should be availability
                        const firstMessages = yield* gateway(peripheral);
                        const firstMessage = Array.from(firstMessages)[0];
                        expect(firstMessage).toEqual(expect.objectContaining({ type: "availability" }));

                        // Subsequent messages should be sensor data
                        const secondMessages = yield* gateway(peripheral);
                        const secondMessage = Array.from(secondMessages)[0];
                        expect(secondMessage).toEqual(expect.objectContaining({ type: "sensor-data" }));
                    })
                )
            ));

        it("should handle unknown ruuvitags normally if unknown ruuvitags are allowed", async () => {
            const unknownPeripheral = {
                ...peripheral,
                uuid: "unknown",
            } as Peripheral;

            const program = Effect.gen(function* () {
                const gateway = makeRuuvitagGateway({ ...ruuviTagSettings, allow_unknown: true });
                // First call to get availability message
                const firstMessages = yield* gateway(unknownPeripheral);

                // Second call to get sensor data
                const secondMessages = yield* gateway(unknownPeripheral);
                return [...Array.from(firstMessages), ...Array.from(secondMessages)];
            });

            const messages = await Effect.runPromise(testWithFreshDeviceRegistry(program));
            expect(messages).toMatchSnapshot();
        });
    });

    describe("getUnavailableDevices()", () => {
        beforeEach(() => {
            jest.useFakeTimers();
            mockedUuid.mockReturnValue("mock-uuid");
            Settings.now = () => new Date("2019-10-10T00:00:00.000Z").valueOf();
        });

        afterEach(() => {
            jest.useRealTimers();
            Settings.now = originalNow;
        });

        it("should produce unavailability messages for ruuvitags that haven't been observed inside timeout", async () => {
            const deviceRegistry = makeRuuvitagDeviceRegistry(ruuviTagSettings, {
                defaultDecimalPrecision: 2,
            });

            const program = Effect.gen(function* () {
                const gateway = makeRuuvitagGateway(ruuviTagSettings);
                const messages = yield* gateway(peripheral);
                expect(Array.from(messages)).toHaveLength(2);
            });

            const TestLayerWithDeviceRegistry = Layer.merge(
                TestLayer,
                Layer.succeed(DeviceRegistryService, deviceRegistry)
            );

            await Effect.runPromise(Effect.provide(program, TestLayerWithDeviceRegistry));

            Settings.now = () => new Date("2019-10-10T00:01:00.000Z").valueOf(); // Advance 1 minute
            const unavailableDevices = deviceRegistry.getUnavailableDevices();
            expect(unavailableDevices).toMatchSnapshot();
        });
    });
});
