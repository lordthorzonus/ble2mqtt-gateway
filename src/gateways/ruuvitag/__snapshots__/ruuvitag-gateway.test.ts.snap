// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuuviTag Gateway getUnavailableDevices() should produce unavailability messages for ruuvitags that haven't been observed inside timeout 1`] = `
[
  {
    "availability": "offline",
    "decimalPrecision": 2,
    "device": {
      "friendlyName": "my_ruuvitag",
      "id": "a1:b2",
      "type": "ruuvitag",
    },
    "lastPublishedAvailability": "online",
    "lastSeen": "2019-10-10T00:00:00.000Z",
    "timeout": 30000,
  },
]
`;

exports[`RuuviTag Gateway ruuvitagGateway() should handle unknown ruuvitags normally if unknown ruuvitags are allowed 1`] = `
[
  {
    "device": {
      "friendlyName": "unknown",
      "id": "unknown",
      "macAddress": "a1:b2",
      "rssi": 23,
      "timeout": 30000,
      "type": "ruuvitag",
    },
    "id": "mock-uuid",
    "payload": {
      "state": "online",
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "availability",
  },
  {
    "device": {
      "friendlyName": "unknown",
      "id": "unknown",
      "macAddress": "CB:B8:33:4C:88:4F",
      "rssi": 23,
      "timeout": 30000,
      "type": "ruuvitag",
    },
    "deviceType": "ruuvitag",
    "id": "mock-uuid",
    "payload": {
      "absoluteHumidity": 11.84,
      "accelerationX": 0,
      "accelerationY": -0,
      "accelerationZ": 1.04,
      "batteryVoltage": 2.98,
      "dewPoint": 14.2,
      "heatIndex": 26,
      "humidex": 28,
      "macAddress": "CB:B8:33:4C:88:4F",
      "measurementSequence": 205,
      "movementCounter": 66,
      "pressure": 100044,
      "relativeHumidityPercentage": 53.49,
      "temperature": 24.3,
      "txPower": 4,
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "sensor-data",
  },
  {
    "device": {
      "friendlyName": "unknown",
      "id": "unknown",
      "macAddress": "CB:B8:33:4C:88:4F",
      "rssi": 23,
      "timeout": 30000,
      "type": "ruuvitag",
    },
    "deviceType": "ruuvitag",
    "id": "mock-uuid",
    "payload": {
      "absoluteHumidity": 11.84,
      "accelerationX": 0,
      "accelerationY": -0,
      "accelerationZ": 1.04,
      "batteryVoltage": 2.98,
      "dewPoint": 14.2,
      "heatIndex": 26,
      "humidex": 28,
      "macAddress": "CB:B8:33:4C:88:4F",
      "measurementSequence": 205,
      "movementCounter": 66,
      "pressure": 100044,
      "relativeHumidityPercentage": 53.49,
      "temperature": 24.3,
      "txPower": 4,
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "sensor-data",
  },
]
`;

exports[`RuuviTag Gateway ruuvitagGateway() should produce a availability and a payload message for a found ruuvitag advertisement 1`] = `
[
  {
    "device": {
      "friendlyName": "my_ruuvitag",
      "id": "a1:b2",
      "macAddress": "a1:b2",
      "rssi": 23,
      "timeout": 30000,
      "type": "ruuvitag",
    },
    "id": "mock-uuid",
    "payload": {
      "state": "online",
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "availability",
  },
  {
    "device": {
      "friendlyName": "my_ruuvitag",
      "id": "a1:b2",
      "macAddress": "CB:B8:33:4C:88:4F",
      "rssi": 23,
      "timeout": 30000,
      "type": "ruuvitag",
    },
    "deviceType": "ruuvitag",
    "id": "mock-uuid",
    "payload": {
      "absoluteHumidity": 11.84,
      "accelerationX": 0,
      "accelerationY": -0,
      "accelerationZ": 1.04,
      "batteryVoltage": 2.98,
      "dewPoint": 14.2,
      "heatIndex": 26,
      "humidex": 28,
      "macAddress": "CB:B8:33:4C:88:4F",
      "measurementSequence": 205,
      "movementCounter": 66,
      "pressure": 100044,
      "relativeHumidityPercentage": 53.49,
      "temperature": 24.3,
      "txPower": 4,
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "sensor-data",
  },
]
`;
